### PR TITLE
server: add tracing to remaining handlers, cleanup code

### PR DESCRIPTION
**Pull request by srenatus**
This deals with the first two bullets of #4128:

1. tracing for remaining handlers
2. decisions IDs are added to the server spans

Also includes some general cleanup.

I'm not sure if that's the convention, but I've put the decision ID into the server spans: the client spans we get from http.send usage in policies will not carry them, but they do refer to their parents, and they'll have the `opa.decision_id` attribute.

Actual usage over time will tell what's proper and useful here, I suppose.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -1420,6 +1425,7 @@ func (s *Server) v1DataGet(w http.ResponseWriter, r *http.Request) {
 			result.Explanation, err = types.NewTraceV1(*buf, pretty)
 			if err != nil {
 				writer.ErrorAuto(w, err)
+				return
```
Comment (srenatus) [inline server/server.go:1428]: From the general code layout, I don't believe this one was omitted with intent.

File: test/e2e/testing.go [hunk 2]
```diff
@@ -50,6 +52,13 @@ func NewAPIServerTestParams() runtime.Params {
 
 	params.GracefulShutdownPeriod = 10 // seconds
 
+	params.DecisionIDFactory = func() string {
+		id, err := uuid.New(rand.Reader)
+		if err != nil {
+			return ""
+		}
+		return id
+	}
```
Comment (srenatus) [inline test/e2e/testing.go:61]: Without this, the e2e test server would not include any decision IDs.

File: server/server.go [hunk 3]
```diff
@@ -951,7 +950,15 @@ func (s *Server) v0QueryPath(w http.ResponseWriter, r *http.Request, urlPath str
 			}
```
Comment (srenatus) [inline server/server.go:950]: ☝️ I've put some of these into the `makeRego` call instead 👇 

Centralizing this, we can change fewer places to enable tracing.

File: server/server.go [hunk 4]
```diff
@@ -91,6 +93,9 @@ const (
 
 const pqMaxCacheSize = 100
 
+// OpenTelemetry attributes
+const otelDecisionIDAttr = "opa.decision_id"
```
Comment (srenatus) [inline server/server.go:97]: ❓ I am not sure if this follows the general conventions; but I've tried 😅

File: server/writer/writer.go [hunk 5]
```diff
@@ -25,32 +25,20 @@ func HTTPStatus(code int) http.HandlerFunc {
...
-	if storage.IsNotFound(err) {
+	case storage.IsNotFound(err):
 		ErrorString(w, http.StatusNotFound, types.CodeResourceNotFound, err)
-		return
+	default:
+		ErrorString(w, http.StatusInternalServerError, types.CodeInternal, err)
 	}
-
-	ErrorString(w, http.StatusInternalServerError, types.CodeInternal, err)
 }
```
Comment (srenatus) [inline server/writer/writer.go:42]: Code cleanup; I prefer the condition-less switch over the if stmt list...

File: server/server.go [hunk 6]
```diff
@@ -1488,18 +1494,20 @@ func (s *Server) v1DataPatch(w http.ResponseWriter, r *http.Request) {
 
 	if err := s.store.Commit(ctx, txn); err != nil {
 		writer.ErrorAuto(w, err)
-	} else {
-		writer.Bytes(w, 204, nil)
+		return
 	}
+
+	writer.Bytes(w, 204, nil)
```
Comment (tsandall) [inline server/server.go:1500]: Nit: use `http.StatusNoContent`

File: server/server.go (hunk 1 above)
Comment (tsandall) [reply server/server.go:1428]: good catch

**Review by tsandall — APPROVED** [review]
LGTM

**Pull request merged**